src/internal/kopia/backup_bases.go

```diff
@@ -0,0 +1,387 @@
+	"github.com/alcionai/corso/src/pkg/logger"
+)
+
+// TODO(ashmrtn): Move this into some inject package. Here to avoid import
+// cycles.
+type BackupBases interface {
+	RemoveMergeBaseByManifestID(manifestID manifest.ID)
+	Backups() []BackupEntry
+	MinBackupVersion() int
+	MergeBases() []ManifestEntry
+	ClearMergeBases()
+	AssistBases() []ManifestEntry
+	ClearAssistBases()
+	MergeBackupBases(
+		ctx context.Context,
+		other BackupBases,
+		reasonToKey func(Reason) string,
+	) BackupBases
+}
+
+type backupBases struct {
+	// backups and mergeBases should be modified together as they relate similar
+	// data.
+	backups     []BackupEntry
+	mergeBases  []ManifestEntry
+	assistBases []ManifestEntry
+}
+
+func (bb *backupBases) RemoveMergeBaseByManifestID(manifestID manifest.ID) {
+	if idx := slices.IndexFunc(
+		bb.mergeBases,
+		func(man ManifestEntry) bool {
+			return man.ID == manifestID
+		},
+	); idx >= 0 {
+		bb.mergeBases = slices.Delete(bb.mergeBases, idx, idx+1)
+	}
+
+	// TODO(ashmrtn): This may not be strictly necessary but is at least easier to
+	// reason about.
+	if idx := slices.IndexFunc(
+		bb.assistBases,
+		func(man ManifestEntry) bool {
+			return man.ID == manifestID
+		},
+	); idx >= 0 {
+		bb.assistBases = slices.Delete(bb.assistBases, idx, idx+1)
+	}
+
+	if idx := slices.IndexFunc(
+		bb.backups,
+		func(bup BackupEntry) bool {
+			return bup.SnapshotID == string(manifestID)
+		},
+	); idx >= 0 {
+		bb.backups = slices.Delete(bb.backups, idx, idx+1)
+	}
+}
+
+func (bb backupBases) Backups() []BackupEntry {
+	return slices.Clone(bb.backups)
+}
+
+func (bb *backupBases) MinBackupVersion() int {
+	min := version.NoBackup
+
+	if bb == nil {
+		return min
+	}
+
+	for _, bup := range bb.backups {
+		if min == version.NoBackup || bup.Version < min {
+			min = bup.Version
+		}
+	}
+
+	return min
+}
+
+func (bb backupBases) MergeBases() []ManifestEntry {
+	return slices.Clone(bb.mergeBases)
+}
+
+func (bb *backupBases) ClearMergeBases() {
+	bb.mergeBases = nil
+	bb.backups = nil
+}
+
+func (bb backupBases) AssistBases() []ManifestEntry {
+	return slices.Clone(bb.assistBases)
+}
+
+func (bb *backupBases) ClearAssistBases() {
+	bb.assistBases = nil
+}
+
+// MergeBackupBases reduces the two BackupBases into a single BackupBase.
+// Assumes the passed in BackupBases represents a prior backup version (across
+// some migration that disrupts lookup), and that the BackupBases used to call
+// this function contains the current version.
+//
+// reasonToKey should be a function that, given a Reason, will produce some
+// string that represents Reason in the context of the merge operation. For
+// example, to merge BackupBases across a ResourceOwner migration, the Reason's
+// service and category can be used as the key.
+//
+// Selection priority, for each reason key generated by reasonsToKey, follows
+// these rules:
+//  1. If the called BackupBases has an entry for a given reaon, ignore the
```

Comment:
"receiver" is a more accurate term here than "called".
```suggestion
//  1. If the receiver BackupBases has an entry for a given reaon, ignore the
```